Initial BE changes for custom expression literals

appleby opened the pull request:
### Description

Closes QUE-720
Closes QUE-721
Closes QUE-775

* ~~update legacy schema for `::FieldOrExpressionDef` to allow `:value` wrapped literals as a top-level expression. (MLv2 schemas already support this).~~
* ~~update `stage-expressions->legacy-MBQL` and `legacy-expression-for-expression-clause` to not unwrap top-level literal `:value`s~~
* ~~add or update tests for `normalize`, `canonicalize`, legacy schema, `lib.convert`~~
* ^ these were merged in milestone 1
* add some basic QP tests for supported DBs
* add order-by to the list of places to check in `should-nest-expressions?`, so that an expression in an order-by clause causes the query to be nested. Fixes the ORDER BY behavior for literal custom expressions.
* add a case for `:value` literals to the `cond` in `metabase.query-processor.middleware.annotate/infer-expression-type`. This was needed to ensure types propagate when "raising" an expression ref into the outer query in `raise-source-query-expression-ref`
* add a new `:expression-literals` driver feature and enable it for postgres, mysql, bigquery, and snowflake.
* add `lib.metadata/database-supports?` helper function to check if the database associated with a query supports a given feature. A similar (private) function was already defined in lib.underlying, so move it into `lib.metadata` to make it more widely available.
* update existing BE tests to check the "real" feature instead of a test-local one
* update `lib.expression/diagnose-expression` to check the `:expression-literals` feature for conditionally enabling/disabling literal expressions.

### How to verify

See included tests, or check out the e2e tests included in the `custom-expressions-literal` branch, which also contains these BE commits and runs some basic e2e tests as well.

### Demo

Basic conversion of top-level literal expression to/from legacy

```Clojure
dev> (as-> (lib/query meta/metadata-provider (meta/table-metadata :orders)) q
       (lib/expression q "foo" 1)
       (lib/filter q (lib/= 1 (lib/expression-ref q "foo")))
       (lib.convert/->legacy-MBQL q))
{:database 29001,
 :type :query,
 :query
 {:source-table 29060,
  :expressions {"foo" [:value 1 nil]},
  :expression-idents {"foo" "IBI9qqb1cO51QT7WVvGcX"},
  :filter [:= 1 [:expression "foo" {:base-type :type/Integer}]]}}

dev> (lib.convert/->pMBQL *1)
{:lib/type :mbql/query,
 :stages
 [{:lib/type :mbql.stage/mbql,
   :source-table 29060,
   :expressions
   [[:value
     {:effective-type :type/Integer,
      :lib/uuid "1912e092-d27b-43dd-839b-47e9cb9f2b33",
      :lib/expression-name "foo",
      :ident "IBI9qqb1cO51QT7WVvGcX"}
     1]],
   :filters
   [[:=
     #:lib{:uuid "a24ee2e4-0979-4c21-aee7-a139c9e21c8a"}
     1
     [:expression {:base-type :type/Integer, :lib/uuid "2c43f3fc-aa09-4269-90de-19e55ced20e5"} "foo"]]]}],
 :database 29001,
 :lib.convert/converted? true}
```

### Checklist

- [x] Tests have been added/updated to cover changes in this PR

appleby committed:
BE changes for :value wrapped expression literals

appleby committed:
Extend BE test with more literal values

appleby committed:
Add float value to qp test for expression literals

appleby committed:
Add Float and Boolean cases to literal-expression-test

appleby committed:
Remove ^:internal from value clause

:value clauses are now generated by the frontend when creating a top-level expression literal and preserved when
converting legacy <-> pMBQL.

appleby committed:
Add some lib.convert tests for literal custom expressions

appleby committed:
Add legacy schema tests for literal custom expressions

appleby committed:
Add normalize / canonicalize tests for literal custom expressions

appleby committed:
Remove vestigial "date" expression from fields clause in query-with-literal-custom-expression-test

appleby committed:
Fix literal-expression-test for cljs

CLJS parses 1.00 as 1 and hence :type/Integer not :type/Float

appleby committed:
Fix js_test expression-clause-<->-legacy-expression-test for wrapped literals

appleby committed:
Move QP tests into expressions_test.clj and add more tests

appleby committed:
Remove commented out :value and update comment in stage-expressions->legacy-MBQL

ranquild committed:
Empty

appleby committed:
Restrict QP literal expressison tests to supported drivers

appleby commented:
Test failure looks related to https://github.com/metabase/metabase/pull/55210

appleby committed:
Merge remote-tracking branch 'origin/master' into expression-literals-integration

appleby committed:
Update boolish->bool to accept BigDecimal inputs

Apparently some DBs return BigDecimal results for boolean expressions.

appleby committed:
Temporarily remove literal-expression-case-test

Needs more debugging

appleby committed:
Move aggregations tests into expression_aggregations_test.clj

ranquild committed:
Add `value` support in the expression editor (#55203)

appleby committed:
Add some basic filter tests to QP expressions_test.clj

ranquild committed:
Merge master

ranquild committed:
Literals in filters tests (#55283)

ranquild committed:
E2E tests for literals in custom columns (#55288)

ranquild reviewed (approved):
@appleby `Filter` schema needs to be fixed as well 
<img width="1280" alt="Screenshot 2025-03-17 at 16 48 49" src="https://github.com/user-attachments/assets/f8bebbb7-5bac-4fbd-9dd3-d2eed5920d6e" />

bshepherdson commented on test/metabase/query_processor_test/expression_aggregations_test.clj:
```diff
@@ -377,3 +378,82 @@
+  (and (driver/database-supports? driver :expressions database)
+       (driver/database-supports? driver :expression-aggregations database)))
+
+;; literal expressions in aggregations not yet supported for these drivers
+(doseq [driver [:mongo :oracle :redshift :sqlite :sqlserver :vertica]]
+  (defmethod driver/database-supports? [driver ::expression-aggregation-literals]
+    [_driver _feature _database]
+    false))
```
Mentioning specific drivers by name in tests is a major code smell.

If these drivers support `:expressions` and `:expression-aggregations` but _not_ `::expression-aggregation-literals`, then it sounds like this is a **new feature** that should have its own flag, set to true on only those drivers that legitimately support it!

appleby replied:
Yeah, agreed. Previously there was TODO comment here about turning these into a real DB feature, but then I backtracked thinking it "shouldn't" be required, but as you say the failing tests are a good indication that it probably is required in practice.

ranquild replied:
How this is going to be used? The FE will allow literals in all cases (as it does now).

appleby replied:
Well current master won't let you create a custom column with a top-level literal value, right?

But actually I have the same question about how this will work outside of tests. I thought the answer was going to be that drivers that don't support a particular feature raise an error if they're asked to compile something they don't support, but I only found one example of that in the mongo driver so far.

ranquild replied:
@appleby `lib/diagnose-expression` could return an error when top-level literals are used for a database that doesn't support them. And, if we do that, we could merge our changes without fixing all the issues at once.

appleby replied:
Probably makes sense to do that anyway, since it's better than letting the user input the expression and fail when the query runs, even if we raise an explicit error.

Up to @bshepherdson if he's ok to merge this to the integration branch without the driver feature change. If so, I can do it in a follow on PR when I'm back next week.

appleby replied:
I've marked it ready for review. If the tests pass and you want merge to unblock other work in the integration branch, go for it. Otherwise, I'll take a look at the driver feature on Monday.

appleby replied:
Fixed in latest updates. Added a new `:expression-literals` driver feature.

bshepherdson commented on test/metabase/query_processor_test/expressions_test.clj:
```diff
@@ -449,6 +450,234 @@
+(def ^:private standard-literal-expression-values
+  (map second (vals standard-literal-expression-defs)))
+
+(defmethod driver/database-supports? [::driver/driver ::expression-literals]
+  [driver _feature database]
+  (driver/database-supports? driver :expressions database))
+
+;; top-level literal expressions not yet supported for these drivers
+(doseq [driver [:mongo :oracle :redshift :sqlite :sqlserver :vertica]]
+  (defmethod driver/database-supports? [driver ::expression-literals]
+    [_driver _feature _database]
+    false))
```
Having to duplicate the list of drivers that are exceptions around this feature is even smellier :(

bshepherdson commented on test/metabase/query_processor_test/expressions_test.clj:
```diff
@@ -449,6 +450,234 @@
+(deftest ^:parallel nested-literal-expression-test
+  (testing "nested literal expression"
+    ;; TODO Fix this test for H2 (QUE-726)
+    (mt/test-drivers (set/difference (mt/normal-drivers-with-feature ::expression-literals :nested-queries)
+                                     #{:h2})
```
nit

```suggestion
    (mt/test-drivers (disj (mt/normal-drivers-with-feature ::expression-literals :nested-queries)
                           :h2)
```

appleby replied:
fixed

bshepherdson commented on test/metabase/query_processor_test/expressions_test.clj:
```diff
@@ -449,6 +450,234 @@
+
+(deftest ^:parallel order-by-literal-expression-test
+  (testing "order-by integer literal expression"
+    ;; ORDER BY a non-negative integer literal is broadly supported (maybe even standard?) and means "order by the Nth
+    ;; column in the select list". Nevertheless, it's questionable whether anyone would actually want to do this.
+    (mt/test-drivers (mt/normal-drivers-with-feature ::expression-literals)
```
I think this is mis-compiling the query, and we shouldn't allow it.

It's breaking the semantics expected by a Metabase `:order-by`, if it's actually an indirection to order by some other column. A Metabase `:order-by` on some expression means "based on this expression"!

appleby replied:
Yeah there is a slack discussion here about whether we should allow order by for literal custom columns. 

https://metaboat.slack.com/archives/C0645JP1W81/p1742317450874479

This is a good point that not only is it questionable, it's actually surprising and probably not what the user intended event if they did try it. What do you think the right point to raise an error would be? In the lib whenever you try to add an order-by on a literal expression?

appleby replied:
Fixed in latest updates. The test now explicitly verifies that ORDER BY of a literal integer expression does not mean "order by the nth column".

bshepherdson reviewed (changes requested):
Holding up my review since I have serious concerns about `:order-by` of a literal number. Also the features slash exceptions in the tests.

ranquild committed:
E2E tests for literal custom columns in other clauses and fixes (#55295)

ranquild committed:
E2E tests for literal custom columns in other clauses and fixes (#55295)

ranquild committed:
Move changes from the FE branch

ranquild committed:
Merge branch 'expression-literals-integration' into expression-literals-initial-be-changes

ranquild committed:
Move changes from the FE branch

ranquild committed:
Merge branch 'expression-literals-integration' into expression-literals-initial-be-changes

ranquild committed:
Move changes from the FE branch

ranquild committed:
Fix filter schema and false = false filter (#55417)

appleby committed:
Remove tests for offset expressions as non-aggregation expressions

These tests were initially added in #42302, but we disabled offset in custom columns in #43224.

Remove these tests so we can re-enabled nest-expressions when expressions appear in an order-by clause.

https://metaboat.slack.com/archives/C0645JP1W81/p1742317450874479

appleby committed:
Add order-by to the list of places to check in should-nest-expressions?

This is to support order-by of custom expressions, in particular literal custom expressions.

https://metaboat.slack.com/archives/C0645JP1W81/p1742317450874479

appleby committed:
Fix malformed order-by clauses

ranquild committed:
Fix date/string literals effective type (#55427)

* Fix date/string literals effective type

* Fix date/string literals effective type

* Fix date/string literals effective type

* Fix date/string literals effective type

* Fix date/string literals effective type

* Fix date/string literals effective type

* Fix date/string literals effective type

* Fix date/string literals effective type

appleby committed:
Merge remote-tracking branch 'origin/expression-literals-integration' into expression-literals-initial-be-changes

appleby committed:
Rename round-trip-expression-literal-test and move it next to existing literal tests in the same file

appleby committed:
Uncomment filter-literal-boolean-expression-with-no-operator-test

The schema was relaxed in bd82763beb, so now the test passes.

appleby committed:
Add round-trip-filter-expression-test

appleby committed:
Add more cases to legacy schema filter-test

appleby committed:
Uncomment filter-literal-expression-with-and-or-test

appleby commented on test/metabase/lib/convert_test.cljc:
```diff
@@ -1060,21 +1191,3 @@
           (is (= query
                  (lib.convert/->legacy-MBQL (lib.convert/->pMBQL query)))))))))
-
-(deftest ^:parallel round-trip-expression-literal-test
```
I just moved and renamed this above to be next to the existing similar roundtrip test already in this branch.

appleby commented on test/metabase/query_processor_test/offset_test.clj:
```diff
@@ -43,106 +43,6 @@
                 [->local-date 2.0 2.0]
                 (qp/process-query query))))))))
 
-(deftest ^:parallel offset-expression-test
```
These tests were failing after re-enabing nesting for expressions in order-by.

Per slack thread [here](https://metaboat.slack.com/archives/C0645JP1W81/p1742409338657409?thread_ts=1742317450.874479&cid=C0645JP1W81), we no longer allow offset outside of aggregations, so removed these tests.

appleby committed:
Remove unused import

appleby committed:
Review feedback: set/difference -> disj

appleby committed:
Merge remote-tracking branch 'origin/master' into expression-literals-initial-be-changes

Conflicts:
	e2e/test/scenarios/custom-column/cc-literals.cy.spec.ts
	e2e/test/scenarios/question-reproductions/reproductions-3.cy.spec.js
	frontend/src/metabase-lib/v1/expressions/diagnostics.ts
	frontend/src/metabase-lib/v1/expressions/pratt/common.ts
	frontend/src/metabase-lib/v1/expressions/pratt/compiler.unit.spec.ts
	frontend/src/metabase-lib/v1/expressions/recursive-parser.js
	frontend/src/metabase-lib/v1/expressions/recursive-parser.unit.spec.ts
	test/metabase/legacy_mbql/schema_test.cljc
	test/metabase/lib/js_test.cljs

appleby committed:
Merge remote-tracking branch 'origin/master' into expression-literals-initial-be-changes

appleby committed:
Add :expression-literals driver feature (#55662)

* Add the :expression-literals database feature

And enabled it for postgres, mysql, bigquery, and snowflake.

These are the tier 1 and tier 2 DBs, minus redshift which is currently failing some tests.

* Add lib.metadata/database-supports? helper

* Check if :expression-literals feature is enabled in diagnose-expression

appleby committed:
Merge branch 'master' into expression-literals-initial-be-changes

appleby committed:
Use standard-literal-expression-values in order-by test expectation

appleby committed:
Add comment about compiler warning to boolish->bool

appleby committed:
Add breakout-by-literal-expression-test

appleby committed:
Add expression-literals-test to nested-queries-test

appleby committed:
Add compound expressions involving literal custom columns to e2e filters tests

appleby committed:
Use lib.schema.common/is-clause? to detect :value expr in diagnose-expression

appleby commented on src/metabase/lib/expression.cljc:
```diff
@@ -534,6 +535,7 @@
           {:message  (i18n/tru "OFFSET is not supported in custom filters")
            :friendly true})
-        (when (= (first expr) :value)
+        (when (and (= (first expr) :value)
```
Slipped my noticed on the first pass, but I guess this should use `is-clause?` or similar. Will push an update.

appleby committed:
Merge remote-tracking branch 'origin/master' into expression-literals-initial-be-changes

Conflicts:
	modules/drivers/snowflake/src/metabase/driver/snowflake.clj

appleby committed:
Merge remote-tracking branch 'origin/master' into expression-literals-initial-be-changes

Conflicts:
	src/metabase/lib/expression.cljc

bshepherdson commented on test/metabase/query_processor_test/expressions_test.clj:
```diff
@@ -449,6 +449,240 @@
+              (mt/run-mbql-query venues
+                {:expressions {"One" [:value 1 nil]}
+                 :fields      [$id $name [:expression "One"]]
+                 :order-by    [[:asc $name]
+                               [:asc [:expression "One"]]]
```
Don't these need to go the other way around for a good test? This is sorting by name and breaking ties with the expression, but there's no tie to break here.

appleby replied:
Yes, good catch! Fixed.

bshepherdson commented on test/metabase/query_processor_test/expressions_test.clj:
```diff
@@ -449,6 +449,240 @@
+                {:expressions standard-literal-expression-defs
+                 :fields      (into [$id] standard-literal-expression-refs)
+                 :order-by    (into [[:asc  $id]]
+                                    (map #(conj [:asc] %) standard-literal-expression-refs))
```
```suggestion
                                    (map #(vector :asc %) standard-literal-expression-refs))
```

appleby committed:
Review feedback: fix order-by-integer-literal-expression-test

appleby committed:
Review feedback: conj -> vector

appleby commented on src/metabase/lib/metadata.cljc:
```diff
@@ -168,6 +168,15 @@
     (mu/disable-enforcement
       (editable-stages? query stages))))
 
+(mu/defn database-supports? :- :boolean
+  "Does `query`s [[database]] support the given `feature`?"
+  [query   :- ::lib.schema/query
+   feature :- :keyword]
+  (-> query
+      database
+      :features
+      (contains? feature)))
+
```
Sure, makes sense. I will do this is in a follow-on PR.
